std/utf.d

```diff
@@ -2420,6 +2469,167 @@ unittest // bugzilla 12923
     }());
 }
 
+/++
+    Returns whether the given range of characters is well-formed Unicode or not.
+
+    Params:
+        range = The range of charaters to validate.
+        index = The index (in code units) where the failed character was (or `0`
+                if the result is `true`).
+  +/
+pragma(inline, true) bool isValidUnicode(R)(R range)
+    if (isInputRange!R && isSomeChar!(ElementType!R) && !isInfinite!R)
+{
+    size_t failedIndex;
+    return _isValidUnicodeImplTrustWrap!false(range, failedIndex);
+}
+
+/// Ditto
+pragma(inline, true) bool isValidUnicode(R)(R range, out size_t failedIndex)
+    if (isInputRange!R && isSomeChar!(ElementType!R) && !isInfinite!R)
+{
+    return _isValidUnicodeImplTrustWrap!true(range, failedIndex);
+}
+
+pragma(inline, true)
+private bool _isValidUnicodeImplTrustWrap(bool getIndex, R)(R range, out size_t failedIndex) @trusted
```

Comment:
As far as I can tell, it is the character array overloads of `decodeImpl()` which are actually, conceptually, `@trusted`. So, we should just wrap `decodeImpl()` to reflect this:
```D
private template decodeImpl(DecodeType decodeType)
{
    private auto impl(bool canIndex, S)(auto ref S str, ref size_t index)
    {
        static if (is(S : const char[]) || (isInputRange!S && is(Unqual!(ElementEncodingType!S) == char)))
        {
            // ...
        } else
        static if (is(S : const wchar[]) || (isInputRange!S && is(Unqual!(ElementEncodingType!S) == wchar)))
        {
            // ...
        } else
        static if (is(S : const dchar[]) || (isInputRange!S && is(Unqual!(ElementEncodingType!S) == dchar)))
        {
            // ...
        }
    }

    pragma(inline, true)
    {
        auto decodeImpl(S)(auto ref S str, ref size_t index) @trusted
            if (isSomeString!S)
        {
            return impl!true(str, index);
        }

        auto decodeImpl(S)(auto ref S str, ref size_t index)
            if (isInputRange!S && isSomeChar!(ElementEncodingType!S))
        {
            return impl!(isRandomAccessRange!S && hasSlicing!S && hasLength!S)(str, index);
        }
    }
}
```